### DISCUSSION TITLE
Nested formular in edit-form

### DISCUSSION THREAD

**Discussion by gwittm** [neuefische/questions#781] [Web] [answered]
at our edit-page we want to be able to delete individual ingredients. To display all ingredients we adapted our useState with "defaultData". Now all ingredients are displayed. But clicking at a delete-button, all ingredients are removed and additionally our AddRecipe-Formular has the error... we need to define "defaultData" but don´t know where. We would be happy if you could help :)

[https://github.com/gwittm/Digital-Recipe-Book/tree/US7-EditRecipe-G%C3%B6khan](url)

**Comment by NikRoe** [comment]
Hi there,

it looks to me like you were very close to the solution!

In your `RecipeForm` component, you need to change the argument you are passing to your `handleDeleteIngredient` function and also the one you are checking within that function.

It needs to be `ingredient._id`.

The filter in your `handleDeleteIngredients` looks like this afterwards:
```js
const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient._id !== ingredientId
    );
```

Your `StyledDeleteIngredientButton` like this:

```js
<StyledDeleteIngredientButton
                  onClick={() => handleDeleteIngredient(ingredient._id)}
                >
                  X
                </StyledDeleteIngredientButton>
```

To make adding a new recipe possible again, you can add a check to your ingredients state to ensure defaultData exists:
Without the check, JavaScript will try to access the ingredients key in defaultData, even it defaultData is undefined. That will lead to an error. You can fix it like this: 

```js
 const [ingredients, setIngredients] = useState(defaultData?.ingredients || [] );
```

I hope this already makes it work on your end again, let me know otherwise ✌️

**Comment by gwittm** [comment]
Thanks a lot for your help.
It improved, but is not yet working correctly:
When adding new ingredients at our edit-page without having old ones, i can not delete individual ingredients. Clicking the x-button deletes all ingredients. 
As soon as i have at least one default ingredient it works correctly. 
At our AddFormular i can only remove all ingredients, but not individual ones. 
Maybe you could give us another hint ;-)

**Reply by NikRoe** [reply]
I will take another look! Thanks for letting me know ✌️

**Reply by NikRoe** [reply, marked as answer]
You need to add an additional check for the ids. 

If there is defaultData, you have access to the `_id`. If you add a new ingredient, you have access to `ingredientId`.

This means your unordered list can look like this:
```js
<ul>
          {ingredients.map((ingredient) => {
            const id = ingredient._id
              ? ingredient._id
              : ingredient.ingredientId;
            return (
              <li key={id}>
                {ingredient.name} {""} {ingredient.amount}
                {ingredient.unit}
                <StyledDeleteIngredientButton
                  onClick={() => handleDeleteIngredient(id)}
                >
                  X
                </StyledDeleteIngredientButton>
              </li>
            );
          })}
        </ul>
```

And your filter like this:

```js
const updatedIngredients = ingredients.filter((ingredient) => {
      const id = ingredient._id ? ingredient._id : ingredient.ingredientId;
      return id !== ingredientId;
    });
```

However, this is only a workaround, since your ingredients loose their ids when being sent to the database. You can fix this by changing what you pass to your onAddIngredient function in your FormularIngredients component to this:

```js
onAddIngredient({
          ingredientID: nanoid(),
          name: event.target.elements.ingredient.value,
          amount: event.target.elements.amount.value,
          unit: event.target.elements.unit.value,
        });
```

the key ingredientID needs to be written exactly the same as you define it in your model.

By making this change, all new ingredients have the ingredientID, which means your list can then look like this:
```js
<ul>
          {ingredients.map((ingredient) => {
            return (
              <li key={ingredient.ingredientID}>
                {ingredient.name} {""} {ingredient.amount}
                {ingredient.unit}
                <StyledDeleteIngredientButton
                  onClick={() =>
                    handleDeleteIngredient(ingredient.ingredientID)
                  }
                >
                  X
                </StyledDeleteIngredientButton>
              </li>
            );
          })}
        </ul>
```

And your filter like this:
```js
const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient.ingredientID !== ingredientId
    );
```

Edit: This will only work for future recipes. Which means that you might need to remove the ingredients from all existing recipes

**Comment by gwittm** [comment]
It´s getting worse... maybe you could visit us to solve the problem and more important: to explain how it works :/
